// users/deactivator.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.id = 'users/deactivator';

exports.models = [
  require('user/models/user')
];

exports.match = ({email}) =>
{
  return /^Undeliverable: /.test(email.subject) && email.from.includes('microsoft');
};

exports.execute = async ({app, module, email}) =>
{
  if (!app.users)
  {
    return;
  }

  const {User} = module;

  const $in = email.to.map(address =>
  {
    const matches = address.match(/<(.*?@.*?)>/);

    return (matches ? matches[1] : address).toLowerCase();
  });

  const users = await User.find({email: {$in}, active: true}).exec();

  if (!users.length)
  {
    return;
  }

  for (const user of users)
  {
    const {email} = user;

    user.email = '';

    await user.save();

    app.info(`Reset user's e-mail due to receiving an undeliverable message.`, {
      module: 'users',
      submodule: 'deactivator',
      userId: user._id.toString(),
      email
    });

    app.broker.publish(`${User.TOPIC_PREFIX}.edited`, {
      model: user,
      user: {id: null, label: 'System'}
    });
  }
};
