// transportOrders/index.js
// Part of <https://miracle.systems/p/walkner-tp> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const setUpRoutes = require('./routes');
const setUpCommands = require('./commands');
const setUpEvents = require('./events');
const setUpPricing = require('./pricing');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  sioId: 'sio',
  currencyRatesId: 'currencyRates/nbp',
  settingsId: 'settings'
};

exports.models = [
  require('./models/transportOrder'),
  require('user/models/user')
];

exports.republishTopics = [
  'transportOrders.added.*.*',
  'transportOrders.edited.*.*',
  'transportOrders.deleted.*.*'
];

exports.recordTopics = {
  blacklist: [
    'transportOrders.added',
    'transportOrders.edited'
  ]
};

exports.userPrivileges = [
  'TRANSPORT_ORDERS:DISPATCHER', 'TRANSPORT_ORDERS:DRIVER', 'TRANSPORT_ORDERS:USER',
  'TRANSPORT_ORDERS:ALL', 'TRANSPORT_ORDERS:DELETE',
];

exports.optionalModules = {
  'mongoose user express': require('./routes'),
  'sio': require('./commands'),
  'mongoose': require('./events'),
  'mongoose sio currencyRates settings': require('./pricing'),
};

exports.start = (app, module) =>
{
  module.hasAccessTo = (user, transportOrder) =>
  {
    if (!user)
    {
      return false;
    }

    const userModule = app[module.config.userId];
    const canViewAll = userModule.isAllowedTo(user, [['TRANSPORT_ORDERS:DISPATCHER'], ['TRANSPORT_ORDERS:ALL']]);

    return canViewAll || transportOrder.users.map(u => String(u)).includes(user._id.toString());
  };
};
