// transportOrders/index.js
// Part of <https://miracle.systems/p/walkner-tp> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');
const setUpRoutes = require('./routes');
const setUpCommands = require('./commands');
const setUpEvents = require('./events');
const setUpPricing = require('./pricing');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  sioId: 'sio',
  currencyRatesId: 'currencyRates/nbp',
  settingsId: 'settings'
};

exports.models = [
  require('./models/transportOrder'),
  require('user/models/user')
];

exports.republishTopics = [
  'transportOrders.added.*.*',
  'transportOrders.edited.*.*',
  'transportOrders.deleted.*.*'
];

exports.recordTopics = {
  blacklist: [
    'transportOrders.added',
    'transportOrders.edited'
  ]
};

exports.start = (app, module) =>
{
  module.hasAccessTo = function hasAccessTo(user, transportOrder)
  {
    if (!user)
    {
      return false;
    }

    var userModule = app[module.config.userId];
    var canViewAll = userModule.isAllowedTo(user, [['TRANSPORT_ORDERS:DISPATCHER'], ['TRANSPORT_ORDERS:ALL']]);

    return canViewAll || _.includes(_.map(transportOrder.users, String), user._id.toString());
  };

  app.onModuleReady(
    [
      module.config.mongooseId,
      module.config.userId,
      module.config.expressId
    ],
    setUpRoutes.bind(null, app, module)
  );

  app.onModuleReady(
    [
      module.config.sioId
    ],
    setUpCommands.bind(null, app, module)
  );

  app.onModuleReady(
    [
      module.config.mongooseId
    ],
    setUpEvents.bind(null, app, module)
  );

  app.onModuleReady(
    [
      module.config.mongooseId,
      module.config.sioId,
      module.config.currencyRatesId,
      module.config.settingsId
    ],
    setUpPricing.bind(null, app, module)
  );
};
